pkg/syncer/metadatasyncer.go

```diff
@@ -1306,6 +1327,169 @@ func calculateVolumeSnapshotReservedForNamespace(ctx context.Context,
+		log.Debugf("calculateSPRReservedForNamespace: There are no StoragePolicyReservation CRs in namespace %q",
+			namespace)
+		return nil, nil
+	}
+
+	storagePolicyIdToReservedMap := make(map[string]*resource.Quantity)
+	for _, storagePolicyReservation := range sprList.Items {
+		log.Debugf("calculateSPRReservedForNamespace: Processing StoragePolicyReservation CR, "+
+			"Name: %q, Namespace: %q", storagePolicyReservation.Name, storagePolicyReservation.Namespace)
+		expectedReservedMap, err := getExpectedReservedCapacityForSPR(ctx, storagePolicyReservation)
+		if err != nil {
+			return nil, err
+		}
+		for storagePolicyId, capacity := range expectedReservedMap {
+			if storagePolicyIdToReservedMap[storagePolicyId] == nil {
+				storagePolicyIdToReservedMap[storagePolicyId] = resource.NewQuantity(0,
+					resource.BinarySI)
+			}
+			storagePolicyIdToReservedMap[storagePolicyId].Add(*capacity)
+		}
+	}
+	return storagePolicyIdToReservedMap, nil
+}
+
+// getExpectedReservedCapacityForSPR gets the expected reserved capacity for given StoragePolicyReservation.
+func getExpectedReservedCapacityForSPR(ctx context.Context,
+	spr storagepolicyv1alpha2.StoragePolicyReservation) (map[string]*resource.Quantity, error) {
+	log := logger.GetLogger(ctx)
+	log.Infof("Fetching expected reserved capacity for StoragePolicyReservation, Name: %q, Namespace: %q",
+		spr.Name, spr.Namespace)
+
+	// Calculate requested capacity by StorageClass from Spec
+	requestedCapacityMap := calculateRequestedCapacityByStorageClass(spr)
+
+	// Calculate approved capacity by StorageClass from Status
+	approvedCapacityMap := calculateApprovedCapacityByStorageClass(spr)
+
+	// Calculate expected reserved capacity by StorageClass based on requested and approved capacity map
+	expectedReservedCapacityMap := calculateExpectedReservedCapacity(ctx, requestedCapacityMap,
+		approvedCapacityMap)
+
+	// Get storageclass to storage policy ID mapping
+	scToStoragePolicyIDMap, err := fetchStorageClassToStoragePolicyMapping(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert StorageClass capacity map to StoragePolicyID capacity map
+	storagePolicyExpectedReservedMap := make(map[string]*resource.Quantity)
+	for storageClassName, expectedReservedCapacity := range expectedReservedCapacityMap {
+		if expectedReservedCapacity.IsZero() {
+			continue
+		}
+
+		log.Infof("Calculated expected reserved capacity for StorageClass: %s, "+
+			"ExpectedReservedCapacity: %s", storageClassName, expectedReservedCapacity.String())
+		// Get storage policy ID from StorageClass
+		if storagePolicyID, ok := scToStoragePolicyIDMap[storageClassName]; ok {
+			// Create storage policy ID to capacity map
+			capacity := expectedReservedCapacity.DeepCopy()
+			storagePolicyExpectedReservedMap[storagePolicyID] = &capacity
+		} else {
+			errStr := "Couldn't find storage policy ID for StorageClass " + storageClassName + " in the map"
+			log.Errorf(errStr)
+			return nil, errors.New(errStr)
```

Comment:
We should log an error and continue. This way we have tried our best to reserve as much capacities as we can.